src/sandstorm/feature-key-tool.c++

```diff
@@ -0,0 +1,196 @@
+                           "print the details.")
+        .callAfterParsing(KJ_BIND_METHOD(*this, doVerify))
+        .build();
+  }
+
+  kj::MainBuilder::Validity doVerify() {
+    kj::Vector<kj::String> text;
+    for (auto& line: splitLines(readAll(STDIN_FILENO))) {
+      auto trimmed = trim(line);
+      if (trimmed.size() > 0 && trimmed[0] != '-') {
+        text.add(kj::mv(trimmed));
+      }
+    }
+
+    auto sign = base64Decode(kj::strArray(text, ""));
+    auto unsign = kj::heapArray<byte>(sign.size());
+    unsigned long long length;
+
+    auto pk = structToBytes(*FeatureKey::SIGNING_KEY, crypto_sign_ed25519_PUBLICKEYBYTES);
+
+    if (crypto_sign_ed25519_open(unsign.begin(), &length, sign.begin(), sign.size(), pk) != 0) {
+      return "signature check failed";
+    }
+
+    kj::ArrayInputStream input(unsign.slice(0, length));
+    capnp::PackedMessageReader reader(input);
+    context.exitInfo(capnp::prettyPrint(reader.getRoot<FeatureKey>()).flatten());
+  }
+
+  kj::MainFunc getKeygenMain() {
+    return kj::MainBuilder(context, "Sandstorm version " SANDSTORM_VERSION,
+                           "Generates a new key and writes it to stdout in binary.")
+        .callAfterParsing(KJ_BIND_METHOD(*this, doKeygen))
+        .build();
+  }
+
+  kj::MainBuilder::Validity doKeygen() {
+    byte random[crypto_sign_ed25519_SEEDBYTES];
+    randombytes(random, sizeof(random));
+    kj::FdOutputStream(STDOUT_FILENO)
+        .write(random, sizeof(random));
+    context.exit();
+  }
+
+  kj::MainFunc getReadkeyMain() {
+    return kj::MainBuilder(context, "Sandstorm version " SANDSTORM_VERSION,
+                           "Reads a key file and writes the public key as capnp text.")
+        .expectArg("<keyfile>", KJ_BIND_METHOD(*this, loadKey))
+        .callAfterParsing(KJ_BIND_METHOD(*this, doReadkey))
+        .build();
+  }
+
+  kj::MainBuilder::Validity doReadkey() {
+    capnp::MallocMessageBuilder builder;
+    auto pk = builder.getRoot<PublicSigningKey>();
+
+    memcpy(structToBytes(kj::cp(pk), crypto_sign_ed25519_PUBLICKEYBYTES), publicKey,
+           crypto_sign_ed25519_PUBLICKEYBYTES);
+    auto msg = kj::str(
+        "(key0 = 0x", kj::hex(pk.getKey0()), ","
+        " key1 = 0x", kj::hex(pk.getKey1()), ","
+        " key2 = 0x", kj::hex(pk.getKey2()), ","
+        " key3 = 0x", kj::hex(pk.getKey3()), ")\n");
+    kj::FdOutputStream(STDOUT_FILENO).write(msg.begin(), msg.size());
+    context.exit();
+  }
+
+private:
+  kj::ProcessContext& context;
+  kj::Vector<kj::StringPtr> importPath;
+  capnp::SchemaParser schemaParser;
+  capnp::ParsedSchema schema;
+
+  byte key[crypto_sign_ed25519_SECRETKEYBYTES];
+  byte publicKey[crypto_sign_ed25519_PUBLICKEYBYTES];
+
+  const byte* structToBytes(capnp::AnyStruct::Reader reader, size_t size) {
+    auto data = reader.getDataSection();
+    KJ_REQUIRE(data.size() == size);
+    return data.begin();
+  }
+
+  byte* structToBytes(capnp::AnyStruct::Builder builder, size_t size) {
```

Comment:
Agreed. Renamed to `getUnderlyingBytes` (both the reader/const and builder/non-const versions). Also renamed in update-tool.c++ which has similar helpers.